src/lib/mockserver.js

```diff
@@ -139,6 +118,47 @@ class MockServer {
             }
         }
     }
+
+    /**
+     * Spawn docker logs until mocks are set up
+     */
+    async _dockerSpawnLogs() {
+        debug(`> docker logs -f ${this.container}`);
+        const proc = spawn('docker', ['logs', '-f', this.container]);
+        let isRunning;
+
+        proc.stdout.on('data', function(data) {
+            const stdout = data.toString();
+            if (stdout && stdout.includes('started on ports: [80, 443]')) {
+                debug("Stdout", stdout);
+                isRunning = true;
+            }
+        });
+        proc.stderr.on('data', function(data) {
+            process.stderr.write(data.toString());
+        });
+
+        const kill = proc.kill;
+        proc.kill = function() {
+            // ensure logging stops immediately
+            proc.stdout.removeAllListeners("data");
+            proc.stderr.removeAllListeners("data");
+            kill.apply(proc);
+        };
+
+        let count = 1;
+        while (count <= 20) {
+            await sleep(500); // to account for cold starts
+            debug('count', count, isRunning);
+            if(isRunning) {
+                debug(`${this.container} is up and running`);
+                proc.kill();
+                return Promise.resolve(true);
+            }
+            count++;
+        }
+        return Promise.reject('Error setting up container');
```

Comment:
I would do `setTimeout()` instead of a loop, much simpler. Just decide on a max timeout time, let's say 10 seconds:

```js
setTimeout(() => {
    // end spawn process
    proc.kill();
    Promise.reject("Error setting up container");
}, 10000);
```

And then above in `if (stdout && stdout.includes('started on ports: [80, 443]'))` do the `Promise.resolve()`.

Reply:
Yeah, I had tried that. The issue is the function does not wait until a promise resolves. To do that you need to add a `return new Promise(resolve, reject)` around the logic.

However, this still doesn't work because `if (stdout && stdout.includes('started on ports: [80, 443]'))` is inside a different function (proc.stdout.on) so resolving the promise inside there does not make the `_dockerSpawnLogs()` resolve.

Reply:
Did you use `resolve()` and `reject()` functions from `new Promise((resolve, reject) => ...` instead of `Promise.reject()`/`Promise.resolve()`? The latter only work in certain cases I think.

Reply:
Here is a sketch of what I think should work:

```js
async _dockerSpawnLogs() {
    return new Promise((resolve, reject) => {
        debug(`> docker logs -f ${this.container}`);
        
        const proc = spawn('docker', ['logs', '-f', this.container]);

        // wait limited time
        setTimeout(() => {
            // end spawned process
            proc.kill();
            reject("Error setting up container");
        }, 10000);

        proc.stdout.on('data', function(data) {
            if (data.toString().includes('started on ports: [80, 443]')) {
                proc.kill();
                resolve();
            }
        });
    });
}
```

Reply:
Based on your other comment, as added protection in case on('data' resolves multiple times, guard it with a boolean flag `ready`:

```js
async _dockerSpawnLogs() {
    return new Promise((resolve, reject) => {
        debug(`> docker logs -f ${this.container}`);
        
        const proc = spawn('docker', ['logs', '-f', this.container]);

        // wait limited time
        setTimeout(() => {
            // end spawned process
            proc.kill();
            reject("Error setting up container");
        }, 10000);

        let ready = false;
        proc.stdout.on('data', function(data) {
            // this log output tells us the server is ready to accept requests
            if (!ready && data.toString().includes('started on ports: [80, 443]')) {
                ready = true;
                proc.kill();
                resolve();
            }
        });
    });
}
```

Reply:
Hi, thanks for your comments above. It worked
Silly me, I had forgotten to add `proc.kill()` inside the proc.stdout.on function and that is why I had multiple instances of the stdouts in the logs. Also the test I was using ran for ~10s so I thought that is was waiting until the setTimeout resolves but turns out the test just actually took around 10s.